rust/parser/mod.rs

```diff
@@ -20,5 +20,598 @@
+    }
+}
+
+fn get_role_player_constraint(ctx: Rc<Role_playerContext>) -> ParserResult<RolePlayerConstraint> {
+    let player = get_var(ctx.player().unwrap().VAR_().unwrap());
+    if let Some(type_) = ctx.type_() {
+        match visit_type(type_)? {
+            Type::Unscoped(label) => Ok(RolePlayerConstraint::from((label, player))),
+            Type::Variable(var) => Ok(RolePlayerConstraint::from((var, player))),
+            _ => Err(ILLEGAL_STATE.format(&[line!().to_string().as_str()])),
+        }
+    } else {
+        Ok(RolePlayerConstraint::from(player))
+    }
+}
+
+fn get_role_players(ctx: Rc<RelationContext>) -> ParserResult<Vec<RolePlayerConstraint>> {
+    (0..)
+        .map_while(|i| ctx.role_player(i))
+        .map(get_role_player_constraint)
+        .collect()
+}
+
+pub fn visit_eof_query(ctx: Rc<Eof_queryContext>) -> ParserResult<Query> {
+    visit_query(ctx.query().unwrap())
+}
+
+fn visit_eof_queries(ctx: Rc<Eof_queriesContext>) -> ParserResult<Vec<Query>> {
+    (0..).map_while(|i| ctx.query(i)).map(visit_query).collect()
+}
+
+fn visit_eof_pattern(ctx: Rc<Eof_patternContext>) -> ParserResult<Pattern> {
+    visit_pattern(ctx.pattern().unwrap())
+}
+
+fn visit_eof_patterns(ctx: Rc<Eof_patternsContext>) -> ParserResult<Vec<Pattern>> {
+    visit_patterns(ctx.patterns().unwrap())
+}
+
+fn visit_eof_definables(ctx: Rc<Eof_definablesContext>) -> ParserResult<Vec<Definable>> {
+    let definables_ctx = ctx.definables().unwrap();
+    (0..)
+        .map_while(|i| definables_ctx.definable(i))
+        .map(visit_definable)
+        .collect()
+}
+
+fn visit_eof_variable(ctx: Rc<Eof_variableContext>) -> ParserResult<Variable> {
+    visit_pattern_variable(ctx.pattern_variable().unwrap())
+}
+
+fn visit_eof_label(ctx: Rc<Eof_labelContext>) -> ParserResult<String> {
+    Ok(ctx.label().unwrap().get_text())
+}
+
+fn visit_eof_schema_rule(ctx: Rc<Eof_schema_ruleContext>) -> ParserResult<()> {
+    visit_schema_rule(ctx.schema_rule().unwrap())
+}
+
+fn visit_query(ctx: Rc<QueryContext>) -> ParserResult<Query> {
+    if let Some(query_match) = ctx.query_match() {
+        Ok(visit_query_match(query_match)?.into_query())
+    } else {
+        Err(ILLEGAL_GRAMMAR.format(&[&ctx.get_text()]))
+    }
+}
+
+fn visit_query_define(_ctx: Rc<Query_defineContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_query_undefine(_ctx: Rc<Query_undefineContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_query_insert(_ctx: Rc<Query_insertContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_query_delete_or_update(_ctx: Rc<Query_delete_or_updateContext>) -> ParserResult<()> {
+    todo!()
+}
+
+fn visit_query_match(ctx: Rc<Query_matchContext>) -> ParserResult<TypeQLMatch> {
+    let mut match_query = typeql_match(visit_patterns(ctx.patterns().unwrap())?).into_match();
+    if let Some(modifiers) = ctx.modifiers() {
+        if let Some(filter) = modifiers.filter() {
```

Comment:
This looks like it would be cleaner with a `match` expression (more similar examples in this same file)

Reply:
Why? This is textbook `if let`
```rust
if let Some(modifiers) = ctx.modifiers() {
    if let Some(filter) = modifiers.filter() {
       // ...
   }
   // ...
}
```
vs
```rust
match ctx.modifiers() {
    Some(modifiers) => {
        match modifiers.filter() {
           Some(filter) => // ...
           None => (),
       }
    }
    None => (),
}
```